CANTroller2/src/TemperatureSensorManager.h

```diff
@@ -0,0 +1,180 @@
+    void assign_remaining_addresses() {
+        auto it = all_locations.begin();
+        for (auto& detected_address : detected_addresses) {
+            if (std::find_if(sensors.begin(), sensors.end(), [&](const std::pair<const sensor_location, TemperatureSensor>& pair) {
+                return std::equal(pair.second.get_address().begin(), pair.second.get_address().end(), detected_address.begin());
+            }) == sensors.end()) {
+                while (it != all_locations.end() && sensors.find(*it) != sensors.end()) {
+                    ++it;
+                }
+                if (it != all_locations.end()) {
+                    // The sensor doesn't exist yet, so create it and add it to the map
+                    sensors.emplace(*it, TemperatureSensor(*it, detected_address, &tempsensebus));
+                    // Print the sensor address for debugging purposes
+                    sensors.at(*it).print_address();
+                }
+            }
+        }
+    }
+
+public:
+    TemperatureSensorManager(uint8_t _onewire_pin) : one_wire_bus(_onewire_pin), tempsensebus(&one_wire_bus) {}
+    void setup() {
+        Serial.println("Setting up Temperature Sensors...");
+        
+        tempsensebus.setWaitForConversion(false);
+        tempsensebus.setCheckForConversion(true);
+        tempsensebus.begin();
+        detected_devices_ct = tempsensebus.getDeviceCount();
+        detected_addresses.resize(detected_devices_ct);
+        Serial.printf (" detected %d devices, parasitic power is %s\n", detected_devices_ct, (tempsensebus.isParasitePowerMode()) ? "on" : "off");  // , DEC);
+        if (detected_devices_ct == 0) {
+            Serial.println("No temperature sensor devices detected");
+            return;
+        }
+
+        // find all detected devices and set their temperature precision
+        for (int i = 0; i < detected_devices_ct; i++) {
+            if (tempsensebus.getAddress(detected_addresses[i].data(), i)) {
+                tempsensebus.setResolution(detected_addresses[i].data(), temperature_precision);
+            }
+        }
+
+        // Assign known addresses to the sensors they belong to
+        assign_known_addresses();
+
+        // Assign remaining addresses to the sensors in order using all_locations
+        assign_remaining_addresses();
+
+        // Request temperature for each sensor, this will make the is_ready() method work
+        request_temperatures();
+    }
+
+    // previously called temp_soren()
+    void update_temperatures() {
+        bool all_sensors_ready = true;
+
+        // Iterate through the sensors that have valid addresses, check if they're ready, and update
+        for (auto& sensor : sensors) {
+            if (sensor.second.has_valid_address()) {
+                if (sensor.second.is_ready()) {
+                    // The sensor is ready, so read the temperature
+                    sensor.second.read_temperature();
```

Comment:
i dunno how long it takes to get a reading, but you could add some state to keep track of if the temp has been read, so that we don't bother re-reading a temp if we don't have to.